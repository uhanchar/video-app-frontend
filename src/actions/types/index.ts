export * from 'actions/types/app-types';
export * from 'actions/types/video-types';
