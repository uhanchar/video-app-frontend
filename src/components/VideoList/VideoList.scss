.video-list {
  display: flex;
  justify-content: center;
  padding: 2rem;
}
