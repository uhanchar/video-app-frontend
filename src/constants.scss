// colors
$color-blank: #ffffff;
$color-success: #2ed47a;
$primary-color-black: #020003;

// sizes
$header-height: 4rem;
