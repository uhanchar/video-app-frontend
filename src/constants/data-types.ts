export type NullableType<T> = T | null;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomAny = any;
