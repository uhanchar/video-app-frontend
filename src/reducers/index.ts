import { combineReducers } from 'redux';
import AppReducer from 'reducers/app.reducer';

const rootReducer = combineReducers({
  app: AppReducer,
});

export default rootReducer;
