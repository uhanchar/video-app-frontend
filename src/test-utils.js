import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { applyMiddleware, createStore } from 'redux';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';

import rootReducer from './reducers';

const reducerInitialState = {};

function render(
  ui,
  {
    initialState = reducerInitialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  } = {},
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      // eslint-disable-next-line react/jsx-filename-extension
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';

export { render };
