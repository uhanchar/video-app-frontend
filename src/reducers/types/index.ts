import { NullableType } from 'constants/data-types';

export interface IVideoItem {
  id: string;
  name: string;
  description: string;
  link: string;
  thumbnailLink: string;
  createdAt: Date;
}

export interface IAppState {
  videos: IVideoItem[];
}

export interface IVideoState {
  video: NullableType<IVideoItem>;
  isLoading: boolean;
  error: string;
}

export type RootState = {
  app: IAppState;
  video: IVideoState;
};
