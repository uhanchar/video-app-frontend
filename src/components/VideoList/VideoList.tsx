import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import 'components/VideoList/VideoList.scss';
import { loadVideos } from 'actions/app.actions';
import { IVideoItem, RootState } from 'reducers/types';
import { Grid } from '@material-ui/core';
import VideoCard from 'components/VideoCard/VideoCard';

const VideoList = () => {
  const dispatch = useDispatch();
  const { videos } = useSelector((state: RootState) => state.app);

  useEffect(() => {
    dispatch(loadVideos());
  }, []);

  return (
    <div className="video-list">
      <Grid container item xs={12} spacing={2}>
        {videos.map((video: IVideoItem) => (
          <Grid item xs={12} sm={6} md={4} lg={2} key={video.id}>
            <VideoCard {...video} />
          </Grid>
        ))}
      </Grid>
    </div>
  );
};

export default VideoList;
