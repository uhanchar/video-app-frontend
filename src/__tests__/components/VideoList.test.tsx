import React from 'react';
import { applyMiddleware, createStore } from 'redux';
import rootReducer from 'reducers';
import axios from 'axios';
import { waitFor } from '@testing-library/react';
import thunk from 'redux-thunk';

import { render as renderWithRedux } from 'test-utils';
import VideoList from 'components/VideoList/VideoList';
import { IVideoItem, RootState } from 'reducers/types';
import { CustomAny } from 'constants/data-types';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

const videoItems: IVideoItem[] = [
  {
    id: '39432ff1585cfb0ed1803ccea2929341',
    name: 'enem2',
    description: '',
    link: 'media/videos/39432ff1585cfb0ed1803ccea2929341',
    thumbnailLink: 'media/thumbnails/39432ff1585cfb0ed1803ccea2929341.png',
    createdAt: new Date(),
  },
  {
    id: '4e4682ecadd48d6bb20b9806b7e522bc',
    name: 'Long',
    description: '',
    link: 'media/videos/4e4682ecadd48d6bb20b9806b7e522bc',
    thumbnailLink: 'media/thumbnails/4e4682ecadd48d6bb20b9806b7e522bc.png',
    createdAt: new Date(),
  },
];

const initialState: RootState = {
  video: {
    video: null,
    isLoading: false,
    error: '',
  },
  app: {
    videos: [],
  },
};

let store: CustomAny;

describe('test for VideoList component', () => {
  beforeEach(() => {
    store = createStore(rootReducer, initialState, applyMiddleware(thunk));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render VideoList component', () => {
    const { asFragment } = renderWithRedux(<VideoList />, { initialState, store });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render proper amount of list items', async () => {
    await mockedAxios.get.mockImplementationOnce(() => Promise.resolve(videoItems));
    const { getAllByTestId } = renderWithRedux(<VideoList />, { initialState, store });

    await waitFor(() => getAllByTestId('video-item'));

    expect(getAllByTestId('video-item').length).toBe(videoItems.length);
  });
});
