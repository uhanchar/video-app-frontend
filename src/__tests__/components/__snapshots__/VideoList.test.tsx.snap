// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for VideoList component should render VideoList component 1`] = `
<DocumentFragment>
  <div
    class="video-list"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-grid-xs-12"
    />
  </div>
</DocumentFragment>
`;
